### PR TITLE
[SMALLFIX] Clean up block writer when failed to cache data read from UFS

### PR DESCRIPTION
**Pull request by apc999**

### CODE REVIEW

File: core/server/worker/src/main/java/alluxio/worker/block/UnderFileSystemBlockReader.java [hunk 1]
```diff
@@ -233,9 +232,8 @@ public int transferTo(ByteBuf buf) throws IOException {
           mBlockWriter.transferFrom(bufCopy);
         }
       } catch (Exception e) {
-        LOG.warn("Failed to cache data read from UFS: %s", e.getMessage());
-        mBlockWriter.close();
-        mBlockWriter = null;
+        LOG.warn("Failed to cache data read from UFS in Alluxio: {}", e.getMessage());
```
Comment (aaudiber) [inline core/server/worker/src/main/java/alluxio/worker/block/UnderFileSystemBlockReader.java]: The two warnings are the same, so it's hard to know which one caused a given log message. Could we make them specific enough to be unique?

File: core/server/worker/src/main/java/alluxio/worker/block/UnderFileSystemBlockReader.java [hunk 2]
```diff
@@ -305,18 +303,16 @@ private void updateUnderFileSystemInputStream(long offset) throws IOException {
...
+    }
     try {
-      if (mBlockWriter != null && offset > mBlockWriter.getPosition()) {
-        mBlockWriter.close();
-        mBlockWriter = null;
-        mLocalBlockStore.abortBlock(mBlockMeta.getSessionId(), mBlockMeta.getBlockId());
-      }
+      mBlockWriter.close();
+      mBlockWriter = null;
+      mLocalBlockStore.abortBlock(mBlockMeta.getSessionId(), mBlockMeta.getBlockId());
```
Comment (aaudiber) [inline core/server/worker/src/main/java/alluxio/worker/block/UnderFileSystemBlockReader.java]: is the exception handling below still appropriate for the new call sites?

File: core/server/worker/src/main/java/alluxio/worker/block/UnderFileSystemBlockReader.java (hunk 1 above)
Comment (calvinjia) [reply core/server/worker/src/main/java/alluxio/worker/block/UnderFileSystemBlockReader.java]: I think we need to request space here too.

Comment (apc999) [reply core/server/worker/src/main/java/alluxio/worker/block/UnderFileSystemBlockReader.java]: done

File: core/server/worker/src/main/java/alluxio/worker/block/UnderFileSystemBlockReader.java (hunk 2 above)
Comment (apc999) [reply core/server/worker/src/main/java/alluxio/worker/block/UnderFileSystemBlockReader.java]: I think so.

File: core/server/worker/src/main/java/alluxio/worker/block/UnderFileSystemBlockReader.java [hunk 3]
```diff
@@ -230,12 +229,13 @@ public int transferTo(ByteBuf buf) throws IOException {
...
         while (bufCopy.readableBytes() > 0) {
+          mLocalBlockStore.requestSpace(mBlockMeta.getSessionId(), mBlockMeta.getBlockId(),
+              mInStreamPos - mBlockWriter.getPosition());
           mBlockWriter.transferFrom(bufCopy);
         }
       } catch (Exception e) {
-        LOG.warn("Failed to cache data read from UFS: %s", e.getMessage());
-        mBlockWriter.close();
-        mBlockWriter = null;
+        LOG.warn("Failed to cache data read from UFS on transferTo(): {}", e.getMessage());
```
Comment (aaudiber) [inline core/server/worker/src/main/java/alluxio/worker/block/UnderFileSystemBlockReader.java]: what does transferTo() mean in terms that a user unfamiliar with the code would understand?

Comment (apc999) [reply core/server/worker/src/main/java/alluxio/worker/block/UnderFileSystemBlockReader.java]: well essentially "read" is the same as "transferTo" but with different argument type.

**Comment by apc999** [comment]
@yupeng9 there actually is an integration test covering this case but not really working (failed to catch the original bug) due to some mockito issue. I fixed the test (and added one). Tested that without my change, the tests will fail.

**Pull request merged**